x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/SecurityIndexRoleMappingCleanupIT.java

```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+package org.elasticsearch.upgrades;
+
+import org.elasticsearch.client.Request;
+import org.elasticsearch.client.RequestOptions;
+import org.elasticsearch.client.Response;
+import org.elasticsearch.client.RestClient;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.xpack.core.security.authc.support.mapper.ExpressionRoleMapping;
+import org.elasticsearch.xpack.core.security.authz.RoleDescriptor;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Set;
+
+import static org.elasticsearch.TransportVersions.SECURITY_ROLE_MAPPINGS_IN_CLUSTER_STATE;
+import static org.elasticsearch.xcontent.XContentFactory.jsonBuilder;
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.Matchers.containsInAnyOrder;
+
+public class SecurityIndexRoleMappingCleanupIT extends AbstractUpgradeTestCase {
+
+    public void testCleanupDuplicateMappingsCreatedWithBug() throws Exception {
+        assumeTrue(
+            "Role mappings are in cluster state after transport version: " + SECURITY_ROLE_MAPPINGS_IN_CLUSTER_STATE,
+            minimumTransportVersion().after(SECURITY_ROLE_MAPPINGS_IN_CLUSTER_STATE)
+        );
+
+        if (CLUSTER_TYPE == ClusterType.OLD) {
+            assumeTrue(
+                "Cleanup only needed before security.role_mapping_cleanup feature available in cluster",
+                clusterHasFeature("security.role_mapping_cleanup") == false
+            );
+            // Since the old cluster has role mappings in cluster state, but doesn't check duplicates, create a duplicate
+            createNativeRoleMapping("operator_role_mapping_1", Map.of("meta", "test"));
+            createNativeRoleMapping("no_name_conflict", Map.of("meta", "test"));
+            assertAllRoleMappings(client(), "no_name_conflict", "operator_role_mapping_1");
+        } else if (CLUSTER_TYPE == ClusterType.UPGRADED) {
+            waitForSecurityMigrationCompletion(adminClient(), 2);
```

Comment:
These tests will fail here until the file service change is merged, since the preconditions for the migration are not met.